# Tugas 2



* Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,
    1. Menggunakan data spam.csv
    2. Fitur CountVectorizer dengan mengaktifkan stop_words
    3. Evaluasi hasilnya

* Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,

    1. Menggunakan data spam.csv
    2. Fitur TF-IDF dengan mengaktifkan stop_words
    3. Evaluasi hasilnya dan bandingkan dengan hasil pada Tugas no 2.
    4. Berikan kesimpulan fitur mana yang terbaik pada kasus data spam.csv



In [8]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
df = pd.read_csv('C:\\kuliah\\Machine Leaning\\2341720003_ML_2025\\data\\spam.csv', encoding='latin-1') # spesifiksi encoding diperlukan karena data tidak menggunakan UTF-8

df.head()

v1                                                 v2 Unnamed: 2  \
0   ham  Go until jurong point, crazy.. Available only ...        NaN   
1   ham                      Ok lar... Joking wif u oni...        NaN   
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3   ham  U dun say so early hor... U c already then say...        NaN   
4   ham  Nah I don't think he goes to usf, he lives aro...        NaN   

  Unnamed: 3 Unnamed: 4  
0        NaN        NaN  
1        NaN        NaN  
2        NaN        NaN  
3        NaN        NaN  
4        NaN        NaN

In [9]:
# Drop 3 kolom terakhir dengan fungsi iloc
df = df.drop(df.iloc[:,2:], axis=1)

# Cek data
df.head()

v1                                                 v2
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

In [10]:
df = df.rename(columns={'v1': 'Labels', 'v2': 'SMS'})

# Cek hasil
print(df.columns)

Index(['Labels', 'SMS'], dtype='object')


In [11]:

# Cek Jumlah Data Per Kelas
print(df['Labels'].value_counts())
print('\n')

# Cek Kelengkapan Data
print(df.info())
print('\n')

# Cek Statistik Deskriptif
print(df.describe())

Labels
ham     4825
spam     747
Name: count, dtype: int64


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5572 entries, 0 to 5571
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Labels  5572 non-null   object
 1   SMS     5572 non-null   object
dtypes: object(2)
memory usage: 87.2+ KB
None


       Labels                     SMS
count    5572                    5572
unique      2                    5169
top       ham  Sorry, I'll call later
freq     4825                      30


In [12]:
# Data untuk label
new_labels = {
    'spam': 1,
    'ham': 0
}

# Encode label
df['Labels'] = df['Labels'].map(new_labels)

# Cek data
df.head()

Labels                                                SMS
0       0  Go until jurong point, crazy.. Available only ...
1       0                      Ok lar... Joking wif u oni...
2       1  Free entry in 2 a wkly comp to win FA Cup fina...
3       0  U dun say so early hor... U c already then say...
4       0  Nah I don't think he goes to usf, he lives aro...

In [13]:
X = df['SMS'].values
y = df['Labels'].values

In [14]:
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

vectorizer_count = CountVectorizer(stop_words='english')
X_train_count = vectorizer_count.fit_transform(X_train)
X_test_count = vectorizer_count.transform(X_test)

# Buat model Naive Bayes
nb_count = MultinomialNB()
nb_count.fit(X_train_count, y_train)

# Prediksi dan evaluasi
y_pred_count = nb_count.predict(X_test_count)

print("=== CountVectorizer ===")
print("Akurasi:", accuracy_score(y_test, y_pred_count))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred_count))
print("Laporan Klasifikasi:\n", classification_report(y_test, y_pred_count))

=== CountVectorizer ===
Akurasi: 0.9832535885167464
Confusion Matrix:
 [[1442   11]
 [  17  202]]
Laporan Klasifikasi:
               precision    recall  f1-score   support

           0       0.99      0.99      0.99      1453
           1       0.95      0.92      0.94       219

    accuracy                           0.98      1672
   macro avg       0.97      0.96      0.96      1672
weighted avg       0.98      0.98      0.98      1672



In [15]:
# Gunakan TF-IDF dengan stop_words
from sklearn.feature_extraction.text import TfidfVectorizer

vectorizer_tfidf = TfidfVectorizer(stop_words='english')
X_train_tfidf = vectorizer_tfidf.fit_transform(X_train)
X_test_tfidf = vectorizer_tfidf.transform(X_test)

# Buat model Naive Bayes
nb_tfidf = MultinomialNB()
nb_tfidf.fit(X_train_tfidf, y_train)

# Prediksi dan evaluasi
y_pred_tfidf = nb_tfidf.predict(X_test_tfidf)

print("=== TF-IDF ===")
print("Akurasi:", accuracy_score(y_test, y_pred_tfidf))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred_tfidf))
print("Laporan Klasifikasi:\n", classification_report(y_test, y_pred_tfidf))


=== TF-IDF ===
Akurasi: 0.9677033492822966
Confusion Matrix:
 [[1453    0]
 [  54  165]]
Laporan Klasifikasi:
               precision    recall  f1-score   support

           0       0.96      1.00      0.98      1453
           1       1.00      0.75      0.86       219

    accuracy                           0.97      1672
   macro avg       0.98      0.88      0.92      1672
weighted avg       0.97      0.97      0.97      1672



### Evaluasi Model Multinomial Naive Bayes

| **Metode**          | **Akurasi** | **Precision (Spam)** | **Recall (Spam)** | **F1-Score (Spam)** | **Keterangan** |
|----------------------|-------------|-----------------------|-------------------|---------------------|----------------|
| **CountVectorizer**  | **0.98**   | 0.95                  | **0.92**          | **0.94**            | Performa terbaik, mampu mengenali pesan spam dan ham dengan seimbang |
| **TF-IDF**           | 0.97       | **1.00**              | 0.75              | 0.86                | Cenderung bias ke kelas “ham”, recall spam lebih rendah |


### Kesimpulan
- **CountVectorizer** menghasilkan performa terbaik dengan akurasi **98%** dan F1-score spam **0.94**.  
- **TF-IDF** memiliki akurasi sedikit lebih rendah (**97%**) dan recall spam hanya **0.75**, menunjukkan banyak spam tidak terdeteksi.  
- **Fitur terbaik:** `CountVectorizer(stop_words='english')` karena hasilnya paling stabil dan akurat untuk mendeteksi spam maupun ham.